#+TITLE:     GPU-based eQTL analysis: Preparing input data
#+AUTHOR:    Kynon J. Benjamin, Ph.D
#+EMAIL:     kynonjade.benjamin@libd.org
#+LANGUAGE:  en
#+HTML_HEAD: <link rel="stylesheet" type="text/css" href="http://gongzhitaao.org/orgcss/org.css"/>
#+PROPERTY:  header-args: :dir /dcs04/lieber/statsgen/jbenjami/tutorials/eqtl_analysis_tutorial
#+PROPERTY:  header-args:R :cache yes :exports both :session *R*
#+PROPERTY:  header-args:python :session *Python* :cache yes :exports both
#+OPTIONS:   H:3 num:nil toc:3 \n:nil @:t ::t |:t ^:{} -:t f:t *:t TeX:t LaTeX:t skip:t d:(HIDE) tags:not-in-toc
#+STARTUP:   align fold nodlcheck hidestars oddeven lognotestate
#+TAGS:      Write(w) Update(u) Fix(f) Check(c) noexport(n)

* eQTL background
Quantitative trait loci (QTL) analysis relays on genetic variant
data (i.e., SNP) and another kind of molecular trait (i.e.,
expression, protein, or DNA methylation). The idea is to test
if a specific molecular trait changes based on allele and
allele dosage.

In general, this pipeline can be used for any kind of quantitative
data with some adaptation. However, I will be focusing on
expression QTL (eQTL) for the remainder of this tutorial.

** Why eQTL analysis
eQTL are genetic variants that explain variation in expression levels.

| [[./img/eqtl_summary.v2.png]] | Translate GWAS to biology                  |
|                           | Useful for gene prioritization             |
|                           | Detection of novel gene-trait associations |
|                           | Inferring direction of association         |

#+begin_src txtimg :results none
#+end_src

Therefore, to test for these associations, we need two main information
from the same samples: 1) expression and 2) genotypes.

* Prepare data
For this analysis, we will be using the BrainSEQ data
([[https://www.ncbi.nlm.nih.gov/pubmed/26687217][PMID: 26687217]]), which saves processed data in R objects.
One important note is that publiclly available human
genotype data is always controlled access due to privacy
issues. This takes time to get access too, so please
address this issue first.

** BrainSEQ data
Processed data are RangedSummarizedExperiment R Objects
(hg38; GENCODE v25).

|-------------------+-----------+-------------+----------------|
| Brain region      | RNA type  | Data        | Reference      |
|-------------------+-----------+-------------+----------------|
| DLPFC             | Poly-A    | [[https://s3.us-east-2.amazonaws.com/jaffe-nat-neuro-2018/rse_gene_BrainSeq_Phase1_hg19_TopHat2_EnsemblV75.rda][Gene counts]] | [[https://pubmed.ncbi.nlm.nih.gov/30050107/][PMID: 30050107]] |
|-------------------+-----------+-------------+----------------|
| DLPFC/Hippocampus | total RNA | [[https://s3.us-east-2.amazonaws.com/libd-brainseq2/rse_gene_unfiltered.Rdata][Gene counts]] | [[https://pubmed.ncbi.nlm.nih.gov/31174959/][PMID: 31174959]] |
|-------------------+-----------+-------------+----------------|
| Caudate nucleus   | total RNA | [[https://caudate-eqtl.s3.us-west-2.amazonaws.com/caudate_brainseq_phase3_hg38_rseGene_merged_n464.rda][Gene counts]] | [[https://pubmed.ncbi.nlm.nih.gov/36319771/][PMID: 36319771]] |
|-------------------+-----------+-------------+----------------|

These object have counts, phenotype information, and
sequencing information that can be used for study
design.

Genotype data is controlled access.

** Study design
Before starting eQTL, it is important to have a goal in
mind. The most common study design is cis-eQTL analysis.
This is assessing variants near (within 1M bp) the transcriptional
start site (TSS). In addition to this, we could also test
for interaction models (genetic variant X phenotype) or
trans-eQTL models. For this tutorial, we will do the more
common cis-eQTL approach. As such, we will use as many
samples as possible to increase our power of detection.

Note: The BrainSEQ data set has neurotypical controls and
individuals with neuropsychiatric disorders (i.e.,
schizophrenia and bipolar disorder). However, diagnosis
status or antipsychotic status does not effect the
eQTL analysis compared with neurotypical controls
([[https://pubmed.ncbi.nlm.nih.gov/36319771/][PMID: 36319771]]).

** Phenotype data

Now lets extract phenotype data for the caudate
nucleus.

#+begin_src sh
  pwd
#+end_src

#+RESULTS:
: /dcs04/lieber/statsgen/jbenjami/tutorials/eqtl_analysis_tutorial

#+begin_src R :results output
  setwd('/dcs04/lieber/statsgen/jbenjami/tutorials/eqtl_analysis_tutorial')
  getwd()
#+end_src

#+RESULTS[bc0abca879696aa55a478e7be0e4830c2c6e58e3]:
: [1] "/dcs04/lieber/statsgen/jbenjami/tutorials/eqtl_analysis_tutorial"

#+BEGIN_SRC R :results output
  caud8_file <- url(paste0("https://caudate-eqtl.s3.us-west-2.amazonaws.com/",
			   "caudate_brainseq_phase3_hg38_rseGene_merged_n464.rda"))
  load(caud8_file)
  rse_gene
#+END_SRC

#+RESULTS[0b4f9366a4fef64c399a09237449b845fd02ff4a]:
: class: RangedSummarizedExperiment 
: dim: 58037 464 
: metadata(0):
: assays(1): counts
: rownames(58037): ENSG00000223972.5 ENSG00000227232.5 ...
:   ENSG00000210195.2 ENSG00000210196.2
: rowData names(10): Length gencodeID ... NumTx gencodeTx
: colnames(464): R12864 R12865 ... R13503 R13504
: colData names(71): BrNum RNum ... RNum.1 FlowCell

#+begin_src R :results silent
					  # Function from jaffelab github
  merge_rse_metrics <- function(rse) {
      stopifnot(is(rse, 'RangedSummarizedExperiment'))

      rse$overallMapRate = mapply(function(r, n) {
	  sum(r*n)/sum(n)
      }, rse$overallMapRate, rse$numReads)
      rse$mitoRate = mapply(function(r, n) {
	  sum(r*n)/sum(n)
      }, rse$mitoRate, rse$numMapped)
      rse$rRNA_rate = mapply(function(r, n) {
	  sum(r*n)/sum(n)
      }, rse$rRNA_rate, rse$numMapped)
      rse$totalAssignedGene = mapply(function(r, n) {
	  sum(r*n)/sum(n)
      }, rse$totalAssignedGene, rse$numMapped)

      rse$numMapped = sapply(rse$numMapped, sum)
      rse$numReads = sapply(rse$numReads, sum)
      rse$numUnmapped = sapply(rse$numUnmapped, sum)
      rse$mitoMapped = sapply(rse$mitoMapped, sum)
      rse$totalMapped = sapply(rse$totalMapped, sum)
      return(rse)
  } 
#+end_src

#+begin_src R :results output
  suppressMessages(library(SummarizedExperiment))
  fields   <- c('BrNum', 'RNum', 'Region', 'RIN', 'Age', 'Sex', 'Race', 
		'Dx', 'mitoRate', 'rRNA_rate', 'overallMapRate')
  rse_gene <- merge_rse_metrics(rse_gene)
  colData(rse_gene)$RIN <- sapply(colData(rse_gene)$RIN,"[",1)
  pheno    <- colData(rse_gene)[,fields]
  dim(pheno)
#+end_src

#+RESULTS[514e70ff196776cb36469e2d0b4e55315b46526b]:
: [1] 464  11

#+begin_src R :results output
  head(pheno, 2)
#+end_src

#+RESULTS[646f036be9e9938ee8572b02d39a9f93832870de]:
: DataFrame with 2 rows and 11 columns
:              BrNum        RNum      Region       RIN       Age         Sex
:        <character> <character> <character> <numeric> <numeric> <character>
: R12864      Br1303      R12864     Caudate       9.6     42.98           F
: R12865      Br1320      R12865     Caudate       9.5     53.12           M
:               Race          Dx  mitoRate   rRNA_rate overallMapRate
:        <character> <character> <numeric>   <numeric>      <numeric>
: R12864          AA      Schizo 0.0326539 8.67516e-05       0.909350
: R12865          AA      Schizo 0.0197874 6.97668e-05       0.873484

As we want to use as many samples as possible, we will only do some
basic filtering for our study design:
  1. Including only individual age > 13, and
  2. Limit to self-identified Black and White Americans

#+begin_src R :results output
  table(pheno$Dx, pheno$Race)
#+end_src

#+RESULTS[09097a2347582062050d3e38e6658a33f12678d0]:
:          
:            AA CAUC
:   Bipolar   4   40
:   Control 134  132
:   Schizo   83   71

#+begin_src R :results output
  pheno <- dplyr::filter(as.data.frame(pheno), Age > 13, Race %in% c("AA", "CAUC"))
  head(pheno, 2)
#+end_src

#+RESULTS[a63228976c64d33ce704d744d767f91071569046]:
:         BrNum   RNum  Region RIN   Age Sex Race     Dx   mitoRate    rRNA_rate
: R12864 Br1303 R12864 Caudate 9.6 42.98   F   AA Schizo 0.03265387 8.675159e-05
: R12865 Br1320 R12865 Caudate 9.5 53.12   M   AA Schizo 0.01978740 6.976684e-05
:        overallMapRate
: R12864      0.9093498
: R12865      0.8734840

#+begin_src R :results output
  dim(pheno)
#+end_src

#+RESULTS[343bdc2a4e439630b955fc4d050dcfde43544b92]:
: [1] 444  11

Now, we'll save this as a text file to work with python.

#+begin_src R :results silent
  as.data.frame(pheno) |>
    write.csv(file = 'data/caudate_phenotypes.csv')
#+end_src

** Normalized counts
We next need normalized counts data. The authors of
tensorQTL and fastQTL transform their counts data using
=edgeR= TMM method.

*** Normalize data
#+begin_example
x <- edgeR::calcNormFactors(x, method="TMM")
#+end_example

They used a helper set of functions to convert this R
function into python. However, since we are already
importing data in R, we can skip the steps of converting
counts and a set of normalized expression and applying
the helper function to transform it to normalized expression
with python.

#+begin_src R :results output
  keepIndex <- which(rse_gene$Age > 13 & rse_gene$Race %in% c("AA", "CAUC"))
  rse_gene  <- rse_gene[, keepIndex]
					  # Clean up sample names if needed
  rownames(colData(rse_gene)) <- sapply(strsplit(rownames(colData(rse_gene)), "_"), "[", 1)
					  # Generate DGE list
  x      <- edgeR::DGEList(counts=assays(rse_gene)$counts[, pheno$RNum], 
			   genes=rowData(rse_gene), samples=pheno)
					  # Filter by expression
  design <- model.matrix(~Race, data=x$samples)
  keep.x <- edgeR::filterByExpr(x, design=design)
  print(paste('There are:', sum(keep.x), 'features left!', sep=' '))
  x      <- x[keep.x, , keep.lib.sizes=FALSE]
					  # Normalize library size
  x      <- edgeR::calcNormFactors(x, method="TMM")
#+end_src

#+RESULTS[8f449c169f9bc8fafab540f6a655da342991776d]:
: [1] "There are: 22465 features left!"

Now, we can save normalized counts data.

#+begin_src R :results output
  cpm <- edgeR::cpm(x)
  cpm[1:5, 1:5]
#+end_src

#+RESULTS[2f07e5a75df789e7ef3036ac67a8b81a3f57ee6f]:
:                       R12864    R12865    R12866    R12867    R12868
: ENSG00000227232.5  2.9236868 2.8017454 2.9691493 2.9180193 4.4492773
: ENSG00000279457.3  3.8088397 3.6661137 3.9180527 5.0163702 3.1866445
: ENSG00000228463.9  2.2531165 4.5304820 2.9997591 4.4262090 2.5252655
: ENSG00000236679.2  0.9119757 1.0432031 0.9489034 1.3770428 0.6613791
: ENSG00000237094.11 0.2145825 0.3874754 0.4897566 0.4590143 1.1423820

#+begin_src R :results silent
  write.table(as.data.frame(cpm),
	      file="data/caudate.normalized_expression.tsv",
	      sep="\t", quote=FALSE)
#+end_src

*** Export gene annotation

#+begin_src R :results output
  genes_to_keep <- rownames(x$genes)
  length(genes_to_keep)
#+end_src

#+RESULTS[e42c0e8d30f99122d7022be20232358faaa23aa2]:
: [1] 22465

#+begin_src R :results output
  head(rowRanges(rse_gene), 2)
#+end_src

#+RESULTS[06417b821e141954675bcd055f8bf3c44fb31003]:
#+begin_example
GRanges object with 2 ranges and 10 metadata columns:
                    seqnames      ranges strand |    Length         gencodeID       ensemblID              gene_type
                       <Rle>   <IRanges>  <Rle> | <integer>       <character>     <character>            <character>
  ENSG00000223972.5     chr1 11869-14409      + |      1735 ENSG00000223972.5 ENSG00000223972 transcribed_unproces..
  ENSG00000227232.5     chr1 14404-29570      - |      1351 ENSG00000227232.5 ENSG00000227232 unprocessed_pseudogene
                         Symbol  EntrezID       Class  meanExprs     NumTx                           gencodeTx
                    <character> <integer> <character>  <numeric> <integer>                     <CharacterList>
  ENSG00000223972.5     DDX11L1     84771       InGen 0.00186396         2 ENST00000456328.2,ENST00000450305.2
  ENSG00000227232.5      WASH7P      <NA>       InGen 1.22336500         1                   ENST00000488147.1
  -------
  seqinfo: 25 sequences from an unspecified genome; no seqlengths
#+end_example

#+begin_src R :results output
  annot <- rowRanges(rse_gene) |> as.data.frame() |>
    tibble::rownames_to_column("gene_id") |>
    dplyr::filter(gene_id %in% genes_to_keep) |>
    dplyr::select(seqnames, start, end, gene_id, strand) |>
    dplyr::mutate(index=gene_id) |>
    tibble::column_to_rownames("index")
  head(annot, 2)
#+end_src

#+RESULTS[d302c22d9ad45314a49b7297237fd1392446d7c7]:
:                   seqnames  start    end           gene_id strand
: ENSG00000227232.5     chr1  14404  29570 ENSG00000227232.5      -
: ENSG00000279457.3     chr1 184923 200322 ENSG00000279457.3      -

#+begin_src R :results silent
  data.table::fwrite(annot, "data/gene.bed",
		     sep='\t', row.names=TRUE)
#+end_src

** Genotypes
We have our genotype data as both VCF and PLINK format.
For this tutorial, I will assume the genotypes are already
quality controlled and in PLINK format (BED/FAM/BIM).

*** Population structure
In addition to having genotypes, we also need information on
population structure. To generate this data, we'll use PLINK
to generate MDS data from pruned data.

#+begin_src sh :results output
  echo "**** Make temporary directory ***"
  mkdir -p tmp
#+end_src

#+RESULTS:
: **** Make temporary directory ***

#+begin_src sh :results output
  module load plink/2.0
  
  echo "**** Prune genotypes ****"
  plink2 --bfile input/TOPMed_LIBD_AA_EA \
	 --indep-pairwise 500kb 0.5 \
	 --out tmp/genotypes
#+end_src

#+RESULTS:
#+begin_example
,**** Prune genotypes ****
PLINK v2.00a3LM 64-bit Intel (17 Dec 2021)     www.cog-genomics.org/plink/2.0/
(C) 2005-2021 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to tmp/genotypes.log.
Options in effect:
  --bfile input/TOPMed_LIBD_AA_EA
  --indep-pairwise 500kb 0.5
  --out tmp/genotypes

Start time: Thu Sep 28 16:55:51 2023
515980 MiB RAM detected; reserving 257990 MiB for main workspace.
Allocated 25827 MiB successfully, after larger attempt(s) failed.
Using up to 48 threads (change this with --threads).
1938 samples (725 females, 1209 males, 4 ambiguous; 1938 founders) loaded from
input/TOPMed_LIBD_AA_EA.fam.
7678274 variants loaded from input/TOPMed_LIBD_AA_EA.bim.
Note: No phenotype data present.
Calculating allele frequencies... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%done.
--indep-pairwise (20 compute threads): 0%16%32%49%65%81%98%6775487/7678274 variants removed.
Writing...Variant lists written to tmp/genotypes.prune.in and tmp/genotypes.prune.out .
End time: Thu Sep 28 17:14:27 2023
#+end_example

#+begin_src sh :results output
  echo "**** Filtered genotypes ****"
  plink2 --bfile input/TOPMed_LIBD_AA_EA \
	 --extract tmp/genotypes.prune.in --make-bed \
	 --maf 0.05 --out tmp/TOPMed_LIBD_AA_EA
#+end_src

#+RESULTS:
#+begin_example
,**** Filtered genotypes ****
PLINK v2.00a3LM 64-bit Intel (2 Mar 2021)      www.cog-genomics.org/plink/2.0/
(C) 2005-2021 Shaun Purcell, Christopher Chang   GNU General Public License v3
Logging to tmp/TOPMed_LIBD_AA_EA.log.
Options in effect:
  --bfile input/TOPMed_LIBD_AA_EA
  --extract tmp/genotypes.prune.in
  --maf 0.05
  --make-bed
  --out tmp/TOPMed_LIBD_AA_EA

Start time: Thu Sep 28 17:24:44 2023
515980 MiB RAM detected; reserving 257990 MiB for main workspace.
Allocated 25827 MiB successfully, after larger attempt(s) failed.
Using up to 48 threads (change this with --threads).
1938 samples (725 females, 1209 males, 4 ambiguous; 1938 founders) loaded from
input/TOPMed_LIBD_AA_EA.fam.
7678274 variants loaded from input/TOPMed_LIBD_AA_EA.bim.
Note: No phenotype data present.
--extract: 902787 variants remaining.
Calculating allele frequencies... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%done.
235754 variants removed due to allele frequency threshold(s)
(--maf/--max-maf/--mac/--max-mac).
667033 variants remaining after main filters.
Writing tmp/TOPMed_LIBD_AA_EA.fam ... done.
Writing tmp/TOPMed_LIBD_AA_EA.bim ... done.
Writing tmp/TOPMed_LIBD_AA_EA.bed ... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%69%70%71%72%73%74%75%76%77%78%79%80%81%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%done.
End time: Thu Sep 28 17:24:58 2023
#+end_example

#+begin_src sh :results output
  echo "**** Run MDS with PLINK ****"
  module load plink/1.90b6.6

  plink --bfile tmp/TOPMed_LIBD_AA_EA --cluster \
	--mds-plot 10 --out input/TOPMed_LIBD_AA_EA
#+end_src

** R session information
#+begin_src R :results output
  Sys.time()
  proc.time()
  options(width = 120)
  sessioninfo::session_info()
#+end_src

#+RESULTS[132693f79a91f59bb1bc590fbbc65d79143e64f7]:
#+begin_example
[1] "2023-09-28 11:31:30 EDT"
    user   system  elapsed 
 132.565    5.303 4388.979
[1m[36m─ Session info ───────────────────────────────────────────────────────────────────────[39m[22m
 [3m[90msetting [39m[23m [3m[90mvalue[39m[23m
 version  R version 4.3.1 Patched (2023-09-26 r85227)
 os       CentOS Linux 7 (Core)
 system   x86_64, linux-gnu
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       US/Eastern
 date     2023-09-28
 pandoc   3.1.1 @ /jhpce/shared/jhpce/core/conda/miniconda3-4.11.0/envs/svnR-4.3.x/bin/pandoc

[1m[36m─ Packages ───────────────────────────────────────────────────────────────────────────[39m[22m
 [3m[90mpackage             [39m[23m [3m[90m*[39m[23m [3m[90mversion  [39m[23m [3m[90mdate (UTC)[39m[23m [3m[90mlib[39m[23m [3m[90msource[39m[23m
 abind                  1.4-5     [90m2016-07-21[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 Biobase              * 2.61.0    [90m2023-04-25[39m [90m[2][39m [90mBioconductor[39m
 BiocGenerics         * 0.47.0    [90m2023-04-25[39m [90m[2][39m [90mBioconductor[39m
 bitops                 1.0-7     [90m2021-04-24[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 cli                    3.6.1     [90m2023-03-23[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 crayon                 1.5.2     [90m2022-09-29[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 DelayedArray           0.27.10   [90m2023-07-28[39m [90m[2][39m [90mBioconductor[39m
 dplyr                  1.1.3     [90m2023-09-03[39m [90m[2][39m [90mCRAN (R 4.3.1)[39m
 edgeR                  3.99.0    [90m2023-09-26[39m [90m[2][39m [90mBioconductor[39m
 fansi                  1.0.4     [90m2023-01-22[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 generics               0.1.3     [90m2022-07-05[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 GenomeInfoDb         * 1.37.4    [90m2023-09-07[39m [90m[2][39m [90mBioconductor[39m
 GenomeInfoDbData       1.2.10    [90m2023-04-11[39m [90m[2][39m [90mBioconductor[39m
 GenomicRanges        * 1.53.1    [90m2023-05-04[39m [90m[2][39m [90mBioconductor[39m
 glue                   1.6.2     [90m2022-02-24[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 IRanges              * 2.35.2    [90m2023-06-22[39m [90m[2][39m [90mBioconductor[39m
 lattice                0.21-8    [90m2023-04-05[39m [90m[3][39m [90mCRAN (R 4.3.1)[39m
 lifecycle              1.0.3     [90m2022-10-07[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 limma                  3.57.8    [90m2023-09-24[39m [90m[2][39m [90mBioconductor[39m
 locfit                 1.5-9.8   [90m2023-06-11[39m [90m[2][39m [90mCRAN (R 4.3.1)[39m
 magrittr               2.0.3     [90m2022-03-30[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 Matrix                 1.6-1.1   [90m2023-09-18[39m [90m[3][39m [90mCRAN (R 4.3.1)[39m
 MatrixGenerics       * 1.13.1    [90m2023-07-25[39m [90m[2][39m [90mBioconductor[39m
 matrixStats          * 1.0.0     [90m2023-06-02[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 pillar                 1.9.0     [90m2023-03-22[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 pkgconfig              2.0.3     [90m2019-09-22[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 R6                     2.5.1     [90m2021-08-19[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 Rcpp                   1.0.11    [90m2023-07-06[39m [90m[2][39m [90mCRAN (R 4.3.1)[39m
 RCurl                  1.98-1.12 [90m2023-03-27[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 rlang                  1.1.1     [90m2023-04-28[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 S4Arrays               1.1.6     [90m2023-08-30[39m [90m[2][39m [90mBioconductor[39m
 S4Vectors            * 0.39.2    [90m2023-09-22[39m [90m[2][39m [90mBioconductor[39m
 sessioninfo            1.2.2     [90m2021-12-06[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 SparseArray            1.1.12    [90m2023-08-31[39m [90m[2][39m [90mBioconductor[39m
 statmod                1.5.0     [90m2023-01-06[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 SummarizedExperiment * 1.31.1    [90m2023-05-01[39m [90m[2][39m [90mBioconductor[39m
 tibble                 3.2.1     [90m2023-03-20[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 tidyselect             1.2.0     [90m2022-10-10[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 utf8                   1.2.3     [90m2023-01-31[39m [90m[2][39m [90mCRAN (R 4.3.0)[39m
 vctrs                  0.6.3     [90m2023-06-14[39m [90m[2][39m [90mCRAN (R 4.3.1)[39m
 XVector                0.41.1    [90m2023-05-03[39m [90m[2][39m [90mBioconductor[39m
 zlibbioc               1.47.0    [90m2023-04-25[39m [90m[2][39m [90mBioconductor[39m

[90m [1] /users/jbenjami/R/4.3.x[39m
[90m [2] /jhpce/shared/jhpce/core/conda/miniconda3-4.11.0/envs/svnR-4.3.x/R/4.3.x/lib64/R/site-library[39m
[90m [3] /jhpce/shared/jhpce/core/conda/miniconda3-4.11.0/envs/svnR-4.3.x/R/4.3.x/lib64/R/library[39m

[1m[36m──────────────────────────────────────────────────────────────────────────────────────[39m[22m
#+end_example

* Pre-process data and align samples
One of the biggest errors I have often run into with using
either fastQTL or tensorQTL is an incorrect order of samples
across expression, genotype, and covariates data. So, this
section focus is getting the input data into a format that
will work with tensorQTL.

** Sample selection and GCT format
This set of functions are used to:
  1. select individuals with genotypes
  2. generate a list to map expression to genotypes IDs (RNum to BrNum)
  3. chromosomes to be assessed
  4. convert normalized counts to GCT format

The GCT format is used by the authors of fastQTL and tensorQTL.
It is not necessary as long as the final input for tensorQTL is in
the right format.

Example script is provided: [[./scripts/01.prepare_gct.py]].

*** Organize data
#+begin_src python :results silent
import pandas as pd
from functools import lru_cache
def to_gct(filename, df):
    description_df = pd.DataFrame({'Description': df.index.values},index=df.index)
    dfo = pd.concat([description_df, df], axis=1)
    dfo.index.name = 'Names'
    with open(filename, "wt") as out:
        print("#1.2", file=out)
        print(df.shape[0], df.shape[1], sep="\t", file=out)
        dfo.to_csv(out, sep="\t")
#+end_src

#+begin_src python :results value
  @lru_cache()
  def get_pheno():
      return pd.read_csv("data/caudate_phenotypes.csv", index_col=0)

  get_pheno().iloc[0:2, 0:6]
#+end_src

#+RESULTS[c56fc1cd4478e8ee5507cc62e8cbffa0e13b1226]:
:          BrNum    RNum   Region  RIN    Age Sex
: R12864  Br1303  R12864  Caudate  9.6  42.98   F
: R12865  Br1320  R12865  Caudate  9.5  53.12   M

#+begin_src python :results value
  @lru_cache()
  def get_fam():
      ## Edit for location of genotypes
      fam_file = "input/TOPMed_LIBD_AA_EA.fam"
      return pd.read_csv(fam_file, sep="\t", header=None,
			 names=["ID","BrNum","V2","V3","V4","V5"])

  get_fam().head(2)
#+end_src

#+RESULTS[04661508400f8fddd2eafb8008edf27665a9daf4]:
:                   ID   BrNum  V2  V3  V4  V5
: 0  3998646007_R01C01  Br2585   0   0   2  -9
: 1  3998646007_R02C01  Br2565   0   0   2  -9

#+begin_src python :results output
  @lru_cache()
  def load_data():
      pheno_df = get_pheno()
      pheno_df["ids"] = pheno_df.RNum
      pheno_df.set_index("ids", inplace=True)
      norm_df = pd.read_csv("data/caudate.normalized_expression.tsv",
			    sep="\t", index_col=0)
      samples = list(set(norm_df.columns).intersection(set(pheno_df["RNum"])))
      return pheno_df.loc[samples,:], norm_df.loc[:,samples]

  pheno_df, norm_df = load_data()
  print(pheno_df.shape)
  print(norm_df.shape)
#+end_src

#+RESULTS[5bc24cf425b5b86092b5c8dfbf72baabfa425cbd]:
: (444, 11)
: (22465, 444)

Now, we'll extract the selected samples.

#+begin_src python :results value
  def select_idv(pheno_df, norm_df):
      samples = list(set(pheno_df.loc[norm_df.columns,:].BrNum)\
		     .intersection(set(get_fam().BrNum)))
      new_fam = get_fam()[(get_fam()["BrNum"].isin(samples))]\
	  .drop_duplicates(subset="BrNum")
      new_fam.to_csv("data/keepFam.txt", sep='\t', index=False, header=False)
      return pheno_df.loc[:, ["RNum", "BrNum"]]\
		     .reset_index().set_index("BrNum")\
		     .loc[new_fam.BrNum].reset_index().set_index("ids")


  new_pheno = select_idv(pheno_df, norm_df)
  new_pheno.head(2)
#+end_src

#+RESULTS[4696bcd4e84f647c5bd86b0dec6a46b89520aac9]:
:          BrNum    RNum
: ids                   
: R12995  Br2585  R12995
: R13019  Br5073  R13019

#+begin_src python :results silent
  to_gct("data/norm.gct", norm_df.loc[:,new_pheno.index])
  new_pheno.loc[:, ["RNum", "BrNum"]]\
	   .to_csv("data/sample_id_to_brnum.tsv", sep="\t", index=False)
  pd.DataFrame({'chr':['chr'+xx for xx in [str(x) for x in range(1,23)]+['X']]})\
    .to_csv('data/vcf_chr_list.txt', header=False, index=None)
#+end_src

*** Python session information
#+begin_src python :results output
  import session_info
  session_info.show()
#+end_src

#+RESULTS[8e034e54755ae8831da19e296898e71ea491e18c]:
: -----
: pandas              1.5.3
: session_info        1.0.0
: -----
: Python 3.10.10 | packaged by conda-forge | (main, Mar 24 2023, 20:08:06) [GCC 11.3.0]
: Linux-3.10.0-1160.el7.x86_64-x86_64-with-glibc2.17
: -----
: Session information updated at 2023-09-28 11:34

** Genotype formatting
Now that we have samples selected and mapping files, we can format our
genotype data. Note: this will be placed in a protected location.

I'll be working on JHPCE for this. This should also order the samples.
#+begin_src sh
  module load plink/2.0
  plink2 --bfile input/TOPMed_LIBD_AA_EA \
	 --keep data/keepFam.txt --make-bed \
	 --out input/protected_data/genotypes

#+end_src

#+RESULTS:
| PLINK                        | v2.00a3LM                          | 64-bit                              | Intel                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | (17                          | Dec      |      2021) | www.cog-genomics.org/plink/2.0/ |             |            |        |      |
| (C)                          | 2005-2021                          | Shaun                               | Purcell,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | Christopher                  | Chang    |        GNU | General                         | Public      | License    | v3     |      |
| Logging                      | to                                 | input/protected_data/genotypes.log. |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |                              |          |            |                                 |             |            |        |      |
| Options                      | in                                 | effect:                             |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |                              |          |            |                                 |             |            |        |      |
| --bfile                      | input/TOPMed_LIBD_AA_EA            |                                     |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |                              |          |            |                                 |             |            |        |      |
| --keep                       | data/keepFam.txt                   |                                     |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |                              |          |            |                                 |             |            |        |      |
| --make-bed                   |                                    |                                     |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |                              |          |            |                                 |             |            |        |      |
| --out                        | input/protected_data/genotypes     |                                     |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |                              |          |            |                                 |             |            |        |      |
|                              |                                    |                                     |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |                              |          |            |                                 |             |            |        |      |
| Start                        | time:                              | Thu                                 | Sep                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | 28                           | 11:29:56 |       2023 |                                 |             |            |        |      |
| 499853                       | MiB                                | RAM                                 | detected;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | reserving                    | 249926   |        MiB | for                             | main        | workspace. |        |      |
| Allocated                    | 7915                               | MiB                                 | successfully,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | after                        | larger   | attempt(s) | failed.                         |             |            |        |      |
| Using                        | up                                 | to                                  | 64                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | threads                      | (change  |       this | with                            | --threads). |            |        |      |
| 1938                         | samples                            | (725                                | females,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 1209                         | males,   |          4 | ambiguous;                      | 1938        | founders)  | loaded | from |
| input/TOPMed_LIBD_AA_EA.fam. |                                    |                                     |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |                              |          |            |                                 |             |            |        |      |
| 7678274                      | variants                           | loaded                              | from                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | input/TOPMed_LIBD_AA_EA.bim. |          |            |                                 |             |            |        |      |
| Note:                        | No                                 | phenotype                           | data                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | present.                     |          |            |                                 |             |            |        |      |
| --keep:                      | 435                                | samples                             | remaining.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |                              |          |            |                                 |             |            |        |      |
| 435                          | samples                            | (141                                | females,                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 294                          | males;   |        435 | founders)                       | remaining   | after      | main   |      |
| filters.                     |                                    |                                     |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |                              |          |            |                                 |             |            |        |      |
| Writing                      | input/protected_data/genotypes.fam | ...                                 | done.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |                              |          |            |                                 |             |            |        |      |
| Writing                      | input/protected_data/genotypes.bim | ...                                 | done.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |                              |          |            |                                 |             |            |        |      |
| Writing                      | input/protected_data/genotypes.bed | ...                                 | 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%done. |                              |          |            |                                 |             |            |        |      |
| End                          | time:                              | Thu                                 | Sep                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | 28                           | 11:30:31 |       2023 |                                 |             |            |        |      |

** Expression formatting
For expression formatting, we need to:
  1. convert to BED format with gene information (i.e., chromosome, start, end)
  2. replace expression ids with genotype ids
  3. compress and index expression file

For this, we will used an adapted version of [[https://github.com/broadinstitute/gtex-pipeline/blob/master/qtl/src/eqtl_prepare_expression.py][eqtl_prepare_expression.py]] from
the fastQTL/tensorQTL authors. Details on how they used this in the
GTEx QTL workflow can be found [[https://github.com/broadinstitute/gtex-pipeline/blob/master/qtl/README.md][here]].

The modified helper script takes the following input:
1. normalized data: GCT format
2. BED file with gene annotation
3. sample ID mapping file
4. chromosomes to analyze

#+begin_src sh
  python3 ./scripts/02.prepare_expression.py --help
#+end_src

#+RESULTS:
| usage:                    | 02.prepare_expression.py  | [-h]         | [-o         | OUTPUT_DIR] |              |      |            |
| [--sample_id_list         | SAMPLE_ID_LIST]           |              |             |             |              |      |            |
| [--feature                | FEATURE]                  | [--bed_file  | BED_FILE]   |             |              |      |            |
| norm_gct                  | sample_participant_lookup |              |             |             |              |      |            |
| vcf_chr_list              | prefix                    |              |             |             |              |      |            |
|                           |                           |              |             |             |              |      |            |
| Generate                  | normalized                | expression   | BED         | files       | for          | eQTL | analyses   |
|                           |                           |              |             |             |              |      |            |
| positional                | arguments:                |              |             |             |              |      |            |
| norm_gct                  | GCT                       | file         | with        | normalized  | expression   |      |            |
| sample_participant_lookup |                           |              |             |             |              |      |            |
| Lookup                    | table                     | linking      | samples     | to          | participants |      |            |
| vcf_chr_list              | List                      | of           | chromosomes | in          | VCF          |      |            |
| prefix                    | Prefix                    | for          | output      | file        | names        |      |            |
|                           |                           |              |             |             |              |      |            |
| options:                  |                           |              |             |             |              |      |            |
| -h,                       | --help                    | show         | this        | help        | message      | and  | exit       |
| -o                        | OUTPUT_DIR,               | --output_dir | OUTPUT_DIR  |             |              |      |            |
| Output                    | directory                 |              |             |             |              |      |            |
| --sample_id_list          | SAMPLE_ID_LIST            |              |             |             |              |      |            |
| File                      | listing                   | sample       | IDs         | to          | include      |      |            |
| --feature                 | FEATURE                   | gene,        | transcript  | or          | exon         |      |            |
| --bed_file                | BED_FILE                  | this         | is          | the         | bed          | file | annotation |

#+begin_src sh
  module load htslib
  module load samtools

  BED="./data/gene.bed"
  python3 ./scripts/02.prepare_expression.py \
	  --feature gene --bed_file $BED -o data/ \
	  ./data/norm.gct ./data/sample_id_to_brnum.tsv \
	  ./data/vcf_chr_list.txt genes

#+end_src

#+RESULTS:
| Loading               | expression | data   |        |       |          |         |        |      |      |
| Map                   | data       |        |        |       |          |         |        |      |      |
| *                     | 22465      | genes. |        |       |          |         |        |      |      |
| bed_template_df.shape | (22465,    | 4)     |        |       |          |         |        |      |      |
| *                     | 22398      | genes  | remain | after | removing | contigs | absent | from | VCF. |
| Writing               | BED        | file   |        |       |          |         |        |      |      |

#+begin_src sh
ls data/genes*
#+end_src

#+RESULTS:
| data/genes.expression.bed.gz     |
| data/genes.expression.bed.gz.tbi |

** Generate covariates
In concurrent with expression and genotype formatting, we
also need to generate covariates for our gene expression
data.

* cis-eQTL analysis with tensorQTL
** Nominal cis-eQTL analysis
** Permutation analysis
** Conditional analysis
** Post hoc
** Fine mapping with SuSiE
